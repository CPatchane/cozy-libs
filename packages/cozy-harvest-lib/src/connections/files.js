const FILES_DOCTYPE = 'io.cozy.files'

/**
 * Creates a directory from a given path
 * @param  {Object}  client CozyClient
 * @param  {string}  path   Directory path
 * @return {Object}         Directory attributes
 */
export const createDirectoryByPath = async (client, path) => {
  const { data } = await client
    .collection(FILES_DOCTYPE)
    .createDirectoryByPath(path)
  return data
}

/**
 * Retrieves a directory from its path
 * @param  {Object}  client CozyClient
 * @param  {string}  path   Directory path
 * @return {Object}        Created io.cozy.files document
 */
export const statDirectoryByPath = async (client, path) => {
  let response
  try {
    response = await client.collection(FILES_DOCTYPE).statByPath(path)
  } catch (error) {
    if (error && error.status === 404) return null
    throw new Error(error.message)
  }

  return response.data
}

export const filesMutations = client => ({
  createDirectoryByPath: createDirectoryByPath.bind(null, client),
  statDirectoryByPath: statDirectoryByPath.bind(null, client)
})

export default filesMutations
