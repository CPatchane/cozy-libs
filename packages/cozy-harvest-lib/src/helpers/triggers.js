const DEFAULT_CRON = '0 0 0 * * 0' // Once a week, sunday at midnight

/**
 * Build trigger attributes given konnector and account
 * @param  {object} konnector
 * @param  {object} account
 * @return {object} created trigger
 */
export const buildAttributes = ({
  account,
  cron = DEFAULT_CRON,
  folder,
  konnector
}) => {
  const message = {
    account: account._id,
    konnector: konnector.slug
  }

  if (folder) {
    message['folder_to_save'] = folder._id
  }

  return {
    type: '@cron',
    arguments: cron,
    worker: 'konnector',
    message
  }
}

const helpers = {
  buildAttributes
}

export default helpers
