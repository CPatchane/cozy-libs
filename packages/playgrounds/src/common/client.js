import CozyClient from 'cozy-client'

export default new CozyClient({
  scope: ['io.cozy.apps', 'io.cozy.konnectors'],
  schema: {
    apps: {
      doctype: 'io.cozy.apps'
    },
    konnectors: {
      doctype: 'io.cozy.konnectors'
    }
  },
  oauth: {
    clientName: 'Example App',
    softwareID: 'io.cozy.example',
    redirectURI: 'http://localhost:1234/auth'
  }
})
