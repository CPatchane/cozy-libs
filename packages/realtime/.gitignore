# NPM
node_modules/
npm-debug.log
yarn-error.log


# Build
dist/

# Coverage
coverage/


# OS generated files
.DS_Store
.DS_Store?
._*
.Spotlight-V100
.Trashes
ehthumbs.db
Thumbs.db
desktop.ini


# Default
# /!\ KEEP THIS SECTION THE LAST ONE
!.gitkeep
