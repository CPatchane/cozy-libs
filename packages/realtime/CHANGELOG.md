# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="1.1.4"></a>
## [1.1.4](https://github.com/cozy/cozy-realtime/compare/cozy-realtime@1.1.3...cozy-realtime@1.1.4) (2018-08-22)




**Note:** Version bump only for package cozy-realtime

<a name="1.1.3"></a>
## [1.1.3](https://github.com/cozy/cozy-realtime/compare/cozy-realtime@1.1.2...cozy-realtime@1.1.3) (2018-08-17)


### Bug Fixes

* **realtime:** Domain parsing ðŸš‘ ([9ffbc26](https://github.com/cozy/cozy-realtime/commit/9ffbc26))




<a name="1.1.2"></a>
## [1.1.2](https://github.com/cozy/cozy-realtime/compare/cozy-realtime@1.1.1...cozy-realtime@1.1.2) (2018-08-09)




**Note:** Version bump only for package cozy-realtime

<a name="1.1.1"></a>
## 1.1.1 (2018-08-08)




**Note:** Version bump only for package cozy-realtime
