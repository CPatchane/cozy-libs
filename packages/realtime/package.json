{
  "name": "cozy-realtime",
  "version": "1.1.0",
  "description": "Realtime interactions with cozy-stack using Websocket",
  "main": "dist/index.js",
  "author": "Cozy Cloud",
  "license": "MIT",
  "scripts": {
    "test": "true",
    "lint": "eslint --fix src/**",
    "build": "babel src -d dist",
    "prepublishOnly": "yarn build"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/cozy/cozy-realtime.git"
  },
  "devDependencies": {
    "babel-cli": "6.26.0",
    "babel-preset-cozy-app": "0.10.0",
    "eslint-config-cozy-app": "0.10.0"
  }
}
