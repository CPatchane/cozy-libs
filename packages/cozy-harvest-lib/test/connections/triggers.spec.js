/* eslint-env jest */
import { triggersMutations } from 'connections/triggers'
import client from 'cozy-client'
import realtime from 'cozy-realtime'

jest.mock('cozy-client', () => ({
  collection: jest.fn().mockReturnValue({
    create: jest.fn(),
    launch: jest.fn()
  }),
  create: jest.fn(),
  client: {
    token: {
      token: '1234abcd'
    },
    uri: 'cozy.tools:8080'
  }
}))

jest.mock('cozy-realtime', () => ({
  subscribe: jest.fn()
}))

const { createTrigger, launchTrigger, waitForLoginSuccess } = triggersMutations(
  client
)

const fixtures = {
  trigger: {
    type: '@cron',
    worker: 'konnector',
    message: {
      account: '8f75c33780e2487fa474e0965521dc32',
      konnector: 'konnectest'
    }
  },
  createdTrigger: {
    _id: '42817ec169d047e68b912c6f7d7564a2',
    _type: 'io.cozy.triggers',
    type: '@cron',
    worker: 'konnector',
    message: {
      account: '8f75c33780e2487fa474e0965521dc32',
      konnector: 'konnectest'
    }
  },
  launchedJob: {
    _type: 'io.cozy.jobs',
    _id: '2794bb7c3cd64712a427ca4454aef238',
    state: 'queued'
  }
}

describe('Trigger mutations', () => {
  beforeAll(() => {
    client
      .collection()
      .create.mockResolvedValue({ data: fixtures.createdTrigger })
    client.collection().launch.mockResolvedValue({ data: fixtures.launchedJob })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  afterAll(() => {
    jest.restoreAllMocks()
  })

  describe('createTrigger', () => {
    it('calls Cozy Client and returns trigger', async () => {
      const result = await createTrigger(fixtures.trigger)
      expect(client.collection().create).toHaveBeenCalledWith(fixtures.trigger)
      expect(result).toEqual(fixtures.createdTrigger)
    })
  })

  describe('launchTrigger', () => {
    it('calls expected endpoint', async () => {
      const result = await launchTrigger(fixtures.trigger)
      expect(client.collection().launch).toHaveBeenCalledWith(fixtures.trigger)
      expect(result).toEqual(fixtures.launchedJob)
    })
  })

  describe('waitForLoginSuccess', () => {
    const shortLoginResponseTime = 50
    const longLoginResponseTime = 150
    const jobSuccessResponseTime = 100

    const updatedJob = {
      state: 'done'
    }

    beforeAll(() => {
      // Mock realtime to respond at 100ms
      realtime.subscribe.mockImplementation(() => ({
        onUpdate: fn => setTimeout(() => fn(updatedJob), jobSuccessResponseTime)
      }))
    })

    afterAll(() => {
      realtime.subscribe.resetMock()
    })

    it('waits for the given delay', async () => {
      const startTime = new Date().getTime()
      await waitForLoginSuccess({}, shortLoginResponseTime)
      const endTime = new Date().getTime()
      expect(endTime - startTime).toBeGreaterThanOrEqual(shortLoginResponseTime)
      expect(endTime - startTime).toBeLessThan(jobSuccessResponseTime)
    })

    it('handle job end before login delay', async () => {
      const startTime = new Date().getTime()
      await waitForLoginSuccess({}, longLoginResponseTime)
      const endTime = new Date().getTime()
      expect(endTime - startTime).toBeGreaterThanOrEqual(jobSuccessResponseTime)
      expect(endTime - startTime).toBeLessThan(longLoginResponseTime)
    })

    it('ignore unfinished job', async () => {
      // Mock realtime to respond at 100ms
      realtime.subscribe.mockImplementation(() => ({
        onUpdate: fn =>
          setTimeout(() => fn({ state: 'queued' }), jobSuccessResponseTime)
      }))
      const startTime = new Date().getTime()
      await waitForLoginSuccess({}, shortLoginResponseTime)
      const endTime = new Date().getTime()
      expect(endTime - startTime).toBeGreaterThanOrEqual(shortLoginResponseTime)
      expect(endTime - startTime).toBeLessThan(jobSuccessResponseTime)
    })
  })
})
