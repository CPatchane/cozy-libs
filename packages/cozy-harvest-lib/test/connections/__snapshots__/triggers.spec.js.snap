// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trigger mutations watchKonnectorJob should return a job watcher with correct properties 1`] = `
KonnectorJobWatcher {
  "_error": null,
  "_runOption": [Function],
  "_succeed": false,
  "client": Object {
    "collection": [MockFunction],
    "create": [MockFunction],
    "options": Object {
      "uri": "cozy.tools:8080",
    },
    "stackClient": Object {
      "token": Object {
        "token": "1234abcd",
      },
    },
  },
  "disableSuccessTimer": [Function],
  "handleError": [Function],
  "handleSuccess": [Function],
  "handleSuccessDelay": [Function],
  "job": Object {
    "source": "timeout",
    "state": "queued",
  },
  "options": Object {
    "expectedSuccessDelay": 8000,
    "onError": [MockFunction],
    "onLoginSuccess": [MockFunction],
    "onSuccess": [MockFunction],
  },
  "successTimer": null,
}
`;
